### DISCUSSION TITLE
Update `process.pid` resource attribute for forked processes (eg. Passenger)

### DISCUSSION THREAD

**Discussion by ykitamura-mdsol** [open-telemetry/opentelemetry-ruby#1409] [Q&A]
Hi, I'm trying to figure out how to update the `process.pid` resource attribute for forked processes:
https://github.com/open-telemetry/opentelemetry-ruby/blob/ad7c3b9c4fcdf2f74705a9ac08310b7067e97162/sdk/lib/opentelemetry/sdk/resources/resource.rb#L58

and it seems I need to update [`@resource` of tracer provider](https://github.com/open-telemetry/opentelemetry-ruby/blob/ad7c3b9c4fcdf2f74705a9ac08310b7067e97162/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb#L42) in all instrumentations.

so I end up patching this method to [OpenTelemetry::Instrumentation::Registry](https://github.com/open-telemetry/opentelemetry-ruby/blob/opentelemetry-sdk/v1.2.0/registry/lib/opentelemetry/instrumentation/registry.rb) but this method is touching instance variables, it doesn't look so nice:

```rb
  def refresh_resource(resource)
    @lock.synchronize do
      @instrumentation.map(&:instance).each do |instrumentation|
        instrumentation.tracer.instance_variable_get(:@tracer_provider)&.instance_variable_set(:@resource, resource)
      end
    end
  end
```

Please advise me if there is a way to update the `process.pid` resource attribute properly.

Thank you

**Comment by ahayworth** [comment]
`TracerProvider#initialize` accepts a `resource` argument, so in theory one could first create a new `TracerProvider` with the desired resource attributes. Then, any tracer you obtain from this custom `TracerProvider` will have the desired resource attributes as well.

It is not straightforward, though - I wonder if maybe this should be updated whenever the SDK detects a fork... maybe @open-telemetry/ruby-approvers and @open-telemetry/ruby-maintainers have ideas?

**Reply by arielvalentin** [reply]
I do not have a simple answer for this. What do other Language SDKs do?

Resources are immutable once the SDK is initialized we are unable to change them. When the process is forked it takes the initialized Resource SDK with it. The only way to change it would be to reset the Resource SDK but we probably do not want to reinitialize the entire SDK to update the `process.pid` resource attribute, especially if it is using custom resources detectors that make API calls (Azure, GCP, etc...).
